iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+
+  if (this->fd <= 0) {
+    Warning("Unable to create temporary file %s, unable to persist hostdb: %d\n", this->tmp_filename.c_str(), this->fd);
+    cont->handleEvent(REFCOUNT_CACHE_EVENT_SYNC, 0);
+    delete this;
+    return EVENT_DONE;
+  }
+
+  // Write out the header
+  int ret = this->write_to_disk((char *)&this->cc->get_header(), sizeof(RefCountCacheHeader));
+  if (ret < 0) {
+    Warning("Error writing cache header to %s: %d", this->tmp_filename.c_str(), ret);
+    cont->handleEvent(REFCOUNT_CACHE_EVENT_SYNC, 0);
+    delete this;
+    return EVENT_DONE;
+  }
+
+  SET_HANDLER((CacheSyncHandler)&RefCountCacheSync::copy_partition);
+  e->schedule_imm(ET_TASK);
+  return EVENT_CONT;
+}
+
+// do the final mv and close of file handle
+template <class C>
+int
+RefCountCacheSync<C>::finalize_sync()
+{
+  // fsync the fd we have
+  int fsync_ret = this->socket_manager.fsync(this->fd);
+  if (fsync_ret != 0) {
+    return fsync_ret;
+  }
+
+  int dir_fd = this->socket_manager.open(this->dirname.c_str(), O_DIRECTORY); // Correct permissions?
+  if (dir_fd <= 0) {
+    return dir_fd;
+  }
+  // move the file
+  int ret = rename(this->tmp_filename.c_str(), this->filename.c_str());
+
+  if (ret != 0) {
+    return ret;
+  }
+
+  // fsync the dir
+  int fsync_dir_ret = this->socket_manager.fsync(dir_fd);
+  if (fsync_dir_ret != 0) {
+    return fsync_dir_ret;
+  }
+
+  int dir_close_ret = this->socket_manager.close(dir_fd);
+  if (dir_close_ret != 0) {
+    return dir_close_ret;
+  }
+
+  int close_ret = this->socket_manager.close(this->fd);
+  if (close_ret != 0) {
+    return close_ret;
+  }
+
+  if (this->rsb) {
+    RecSetRawStatCount(this->rsb, refcountcache_last_sync_time, Thread::get_hrtime() / HRTIME_SECOND);
+    RecSetRawStatCount(this->rsb, refcountcache_last_total_items, this->total_items);
+    RecSetRawStatCount(this->rsb, refcountcache_last_total_size, this->total_size);
+  }
+
+  return 0;
+}
+
+// Write *i to this->fd, if there is an error we'll just stop this continuation
+// TODO: reschedule the continuation if the disk was busy?
+template <class C>
+int
+RefCountCacheSync<C>::write_to_disk(char *i, int size)
+{
+  int written = 0;
+  while (written < size) {
+    int ret = this->socket_manager.write(this->fd, i + written, size - written);
+    if (ret <= 0) {
+      return -1;
+    } else {
+      written += ret;
+    }
+  }
+  return 0;
+}
+
+template <class C>
+RefCountCacheSync<C>::RefCountCacheSync(Continuation *acont, C *cc, int frequency, std::string dirname, std::string filename)
+  : Continuation(NULL),
+    partition(0),
+    cc(cc),
+    cont(acont),
+    fd(0),
+    dirname(dirname),
+    filename(filename),
+    time_per_partition(HRTIME_SECONDS(frequency) / cc->partition_count()),
+    start(Thread::get_hrtime()),
+    total_items(0),
+    total_size(0),
+    rsb(cc->get_rsb())
+
+{
+  eventProcessor.schedule_imm(this, ET_TASK);
+  mutex              = cc->get_partition(partition).lock.get();
```

Comment:
This locking is much too clever and you are going to end up holding the first partition lock while doing file IO. Better to keep it simple and just acquire the partition lock in ``copy_partition``.